### ISSUE TITLE
MarkDown markup visible on documentation search page

### ISSUE THREAD

**Issue by jmsmkn** [comic/grand-challenge.org#3575]
This should only show the content, not the markup:

<img width="783" alt="Screenshot 2024-10-04 at 20 22 33" src="https://github.com/user-attachments/assets/56aea4ac-50a4-45c6-ab45-0bce0caf9a66">

**Comment by ammar257ammar** [comment]
I was looking into this issue. 
The headline appearing in the search results in each entry is obtained using the function [SearchHeadline()](https://github.com/comic/grand-challenge.org/blob/main/app/grandchallenge/documentation/views.py#L47)
```python
headline = SearchHeadline("content", query)
```
This function acts on the database where it runs the postgresql [ts_headline](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-HEADLINE) function.

The first argument is either the column name or a database expression applied to the column (using `F() `functions).

Therefore, the `SearchHeadline()` will extract the relevant text directly from the stored text which is in markdown format, leading to a stripped markdown (malformed). 

In this case, the `markdown()` function does not work correctly to render the markdown. 
Example:
```python
    md = ("bold**: "
        "more text more text more text. here is the link (defined by [woooow](https://www.google.com")
        
    html = render_markdown(md or "")
    soup = BeautifulSoup(html, features='html.parser')

    # Output: bold**: more text more text more text. here is the link (defined by [woooow](https://www.google.com
```

Ideally, we want the content to be in plain text (e.g. by rendering the markdown to html and then extracting the text using `bs4`) **before** getting the headline from it, but as I see now that is not possible with the function `SearchHeadline()`.

I can think of these solutions currently:
1- remove the markdown characters from the headline text (with regex). That will require handling quite some cases to make sure it is clean.
2- Render `content`, get plain text and use the first `n` character as headline. This will not contain highlighted text in relevance to the query but a snippet from the documentation page. 
3- Render and extract the text from `content`, then do the search with regex and extract the match (first occurence) with surrounding text (e.g. +/- 30 characters) and use that as headline. This solution will not use the `SearchHeadline()`.

**Comment by amickan** [comment]
> but as I see now that is not possible with the function `SearchHeadline()`.

Why is this not possible? I would have thought that you can annotate the queryset with a new `plain_text` field and then pass that to SearchHeadline?

**Comment by amickan** [comment]
I'm not sure, but I think you can write your own database expression using [Func()](https://docs.djangoproject.com/en/5.1/ref/models/expressions/#func-expressions) (and then maybe the `regexp_replace` function of PostGresSQL). Not sure how performant it is, and how hard it would be to get the regex right, but I think this should at least technically be possible. You would then annotate the queryset with a `clean_content` and pass that the `SearchHeadline`.

**Comment by ammar257ammar** [comment]
okay, I believe then that you recommend the first suggestion that removes Markdown characters from the headline.

> 1- remove the markdown characters from the headline text (with regex). That will require handling quite some cases to make sure it is clean.

I will work on this approach.

**Comment by jmsmkn** [comment]
Yes just remove the Headline search, do not implement keyword highlighting or try to fix #3740.
